### PR TITLE
Add util functions for hertz

### PR DESCRIPTION
**Pull request by MaliaLabor**
Issue #230

### CODE REVIEW

File: include/libhal/units.hpp [hunk 1]
```diff
@@ -111,5 +112,93 @@ namespace experimental::literals {
...
+/**
+ * @brief Calculate the number of cycles of this frequency within the time
+ * duration. This function is meant for timers to determine how many count
+ * cycles are needed to reach a particular time duration at this frequency.
+ *
+ * @param p_source - source frequency
+ * @param p_duration - the amount of time to convert to cycles
+ * @return std::int64_t - number of cycles
+ */
+[[nodiscard]] constexpr std::int64_t cycles_per(
```
Comment (kammce) [inline include/libhal/units.hpp:125]: Overall do not use fixed point math here. We should be avoiding division by int64_t.

File: include/libhal/units.hpp [hunk 2]
```diff
@@ -111,5 +112,93 @@ namespace experimental::literals {
...
+/**
+ * @brief Calculate the amount of time it takes a frequency to oscillate a
+ * number of cycles.
+ *
+ * @param p_source - the frequency to compute the cycles from
+ * @param p_cycles - number of cycles within the time duration
+ * @return std::chrono::nanoseconds - time duration based on this frequency
+ * and the number of cycles
+ */
+[[nodiscard]] constexpr std::chrono::nanoseconds duration_from_cycles(
```
Comment (kammce) [inline include/libhal/units.hpp]: Overall do not use fixed point math here. We should be avoiding division by int64_t.

File: include/libhal/units.hpp [hunk 3]
```diff
@@ -111,5 +112,93 @@ namespace experimental::literals {
...
+/**
+ * @brief Calculates and returns the wavelength in seconds.
+ *
+ * @tparam Period - desired period (defaults to std::femto for femtoseconds).
+ * @param p_source - source frequency to convert to wavelength
+ * @return std::chrono::duration<int64_t, Period> - time based wavelength of
+ * the frequency.
+ */
+template<typename Period = std::femto>
+constexpr std::chrono::duration<int64_t, Period> wavelength(hertz p_source)
```
Comment (kammce) [inline include/libhal/units.hpp:152]: Add another function that returns float rather than std::chrono::duration.

File: include/libhal/units.hpp [hunk 4]
```diff
@@ -111,5 +112,93 @@ namespace experimental::literals {
...
+  //                              \ ratio_den /
+  //
+  // std::chrono::nanoseconds::period::num == 1
+  // std::chrono::nanoseconds::period::den == 1,000,000
+
+  constexpr std::int64_t numerator = decltype(p_duration)::period::num;
+  constexpr std::int64_t denominator = decltype(p_duration)::period::den;
+  // Storing 32-bit value in a std::uint64_t for later computation, no
+  // truncation possible.
+  const std::int64_t duration = absolute_value(p_duration.count());
```
Comment (kammce) [inline include/libhal/units.hpp]: You don't need to use absolute value here.

File: include/libhal/units.hpp [hunk 5]
```diff
@@ -111,5 +112,93 @@ namespace experimental::literals {
...
+  // Duration contains at most a 64-bit number, value() is a
+  // 32-bit number, and numerator is always the value 1.
+  //
+  // To contain the maximum resultant possible requires storage within an
+  // integer of size 64-bits, and thus will fit within a std::uint64_t
+  // variable, no overflow checks required.
+  const std::int64_t count =
```
Comment (kammce) [inline include/libhal/units.hpp]: Storing this in int64_t results in us doing integral math when we should be using floating point math.

**Review by kammce — CHANGES_REQUESTED** [review]

**Comment by MaliaLabor** [comment]
I'm not sure if I correctly resolved these issues so I left them unresolved for now

File: include/libhal/units.hpp [hunk 6]
```diff
@@ -111,5 +112,96 @@ namespace experimental::literals {
...
+  //
+  //                /    cycles * ratio_den    \_
+  //   |period| =  | ---------------------------|
+  //                \ frequency_hz * ratio_num /
+  //
+  constexpr auto time_duration_den = std::chrono::nanoseconds::period::den;
+  constexpr auto time_duration_num = std::chrono::nanoseconds::period::num;
+  config::float_type ratio = time_duration_den / time_duration_num;
+  auto nanoseconds = (p_cycles / p_source) * ratio;
+
```
Comment (kammce) [inline include/libhal/units.hpp:208]: Check bounds and if it exceeds int64_t return `std::errc::result_out_of_range` found here: https://en.cppreference.com/w/cpp/error/errc

**Review by kammce — CHANGES_REQUESTED** [review]

File: include/libhal/units.hpp [hunk 7]
```diff
@@ -111,5 +113,103 @@ namespace experimental::literals {
...
+  constexpr auto time_duration_den = std::chrono::nanoseconds::period::den;
+  constexpr auto time_duration_num = std::chrono::nanoseconds::period::num;
+  constexpr auto int_max =
+    static_cast<config::float_type>(std::numeric_limits<std::int64_t>::max());
+  constexpr auto int_min =
+    static_cast<config::float_type>(std::numeric_limits<std::int64_t>::min());
+  config::float_type ratio = time_duration_den / time_duration_num;
+  auto nanoseconds = (p_cycles / p_source) * ratio;
+
+  if (int_min > nanoseconds || nanoseconds > int_max) {
```
Comment (kammce) [inline include/libhal/units.hpp]: This is in the wrong for int_min.

Should be `int_min <= nanoseconds && nanoseconds <= int_max`

**Review by kammce — CHANGES_REQUESTED** [review]

File: include/libhal/units.hpp (hunk 7 above)
Comment (MaliaLabor) [reply include/libhal/units.hpp]: I had to use `if (!(int_min <= nanoseconds && nanoseconds <= int_max))` to get this work, is this correct?

**Review by kammce — APPROVED** [review]

**Pull request merged**